File: R/smooth_gauss.R



#' @title Smooth per-vertex data based on a mesh using true Gaussian smoothing.
#'
#' @description This function performs Gaussian smoothing of vertices in the geodesic neighborhood of all vertices. This is computationally quite expensive and requires large amounts of memory for large meshes.
#'
#' @inheritParams pervertexdata.smoothnn
#'
#' @param fwhm scalar double, smoothing kernel full width at half max
#'
#' @param trunc_factor scalar double, truncation factor for Gaussian neighborhood, in Gaussian standard deviations. Must be greater than 1.0 to make any sense, leave as is if in doubt. Larger values mean that a larger geodesic neighborhood needs to be computed on the mesh, which is slower.
#'
#' @return numerical vector, the smoothed data.
#'
#' @seealso \code{\link{pervertexdata.smoothnn}} can be used to approximate Gaussian smoothing with several iterations of nearest neighbor smoothing, and is a lot faster for large meshes.
#'
#' @examples
#' \dontrun{
#' mesh = rgl::tetrahedron3d();
#' pvd = rnorm(nrow(mes2$vb), mean = 5.0, sd = 1.0);
#' pvd_smoothed = pervertexdata.smoothgauss(mesh, pvd, fwhm=5.0);
#' }
#'
#' @export
pervertexdata.smoothgauss <- function(surface, pvdata, fwhm, trunc_factor=3.5) {

  if(requireNamespace("Rvcg", quietly = TRUE)) {
    return(Rvcg::vcgSmoothPVD(ensure.tmesh3d(surface), pvdata, fwhm = fwhm, trunc_factor = trunc_factor));
  } else {
    stop("The 'Rvcg' package is required to use this functionality. Use 'remotes::install_github('dfsp-spirit/Rvcg', ref='smooth_pervertex_data')' to obtain the correct package version.");
  }
}

File: src/smooth_gauss.cpp

#include <Rcpp.h>
using namespace Rcpp;
// [[Rcpp::plugins(cpp11)]]

#define _USE_MATH_DEFINES
#include <cmath>
#include <vector>
#include <cassert>
//#include <iostream>

/// Perform nearest-neighbor smoothing of the given data, based on mesh adjacency list representation.
/// @param _mesh the input mesh
/// @param _data an R numerical vector with one value per mesh vertex
/// @param _fwhm the FWHM for the Gaussian kernel
/// @param _truncfactor the cutoff factor after which to end the Gaussian neighborhood, in Gaussian standard deviations
RcppExport SEXP smooth_data_gaussian(SEXP _mesh, SEXP _data, SEXP _fwhm, SEXP _truncfactor) {
  return wrap(1);
}

/// Used in C++ code only, not exported.
std::vector<std::vector<float>> gauss_weights(const std::vector<std::vector<int>> geod_neigh_indices, const std::vector<std::vector<float>> geod_neigh_dists, const float gstd) {
  std::vector<std::vector<float>> weights;

  float gvar2 = 2 * (gstd * gstd);
  float f = 1.0 / (sqrt(2 * M_PI) * gstd);

  return(weights);
}

/// Used in C++ code only, not exported.
std::vector<float> spatial_filter(const std::vector<float> data, const std::vector<std::vector<int>> geod_neigh_indices, const std::vector<std::vector<float>> geod_neigh_gauss_weights) {
  std::vector<float> smoothed_data(data.size());
  float smoothed_val;
  for(size_t i=0; i<data.size(); i++) {
    smoothed_val = 0.0;
    for(size_t j=0; j<geod_neigh_indices[i].size(); j++) {
      smoothed_val += data[geod_neigh_indices[i][j]] * geod_neigh_gauss_weights[i][j];
    }
    smoothed_data[i] = smoothed_val;
  }
  return(smoothed_data);
}

/// Used in C++ code only, not exported.
inline float fhwm_to_gstd(const float fwhm) {
  return(fwhm / sqrt(log(256.0)));
}
